### ISSUE TITLE
Selecting PhysX meshes for an .fbx asset in FBX Settings (PREVIEW) when Editor is initially launched crashes the Editor

### ISSUE THREAD

**Issue by LB-MateuszSzweda** [o3de/o3de#6027]
**Describe the bug**
Selecting PhysX meshes for an .fbx asset in FBX Settings (PREVIEW) crashes the Editor upon clicking the _select_ button.

Please note that this occurs only once for each built project and will occur only on the Editor's first launch. The issue cannot be reproduced after restarting the Editor or if the Editor was closed and restarted after building.

What was also observed is that the icons of .fbx assets in the Asset Browser have a placeholder papersheet icon during the initial launch of the Editor, even when Asset Processor processed all the files successfully. Correct icons will be displayed only after restarting the Editor.

The issue occurs for new projects and the AutomatedTesting project.

PC Development build (3922280) and Linux Development (3922280) and Linux Stabilization/2110(6480041) are also affected.

For further information please refer to the attached video file and crash logs.

**Prerequisites**

* o3de was installed by the installer or obtained and built by using instructions from the
https://o3de.org/docs/welcome-guide/setup/setup-from-github/building-windows/ site.
* New project was created and built in Project Manager.
 

**Steps to reproduce**

1. Launch the Editor and create a new level.
2. Search for any _.fbx_ object in the **Asset Browser** (ex. groundplane_512x512m.fbx) .
3. Right click the _.fbx_ asset and select _Edit Settings..._ .
4. Navigate to the _PhysX_ tab and click the menu icon from the _Select meshes_ row.
5. Select all nodes in the opened menu and click the _Select_ button

**Expected behavior**
PhysX meshes are selected for .fbx asset and the Editor remains stable.

**Actual behavior**
The Editor crashes.

**Screenshots/Video**

https://user-images.githubusercontent.com/86953456/144066320-bc3249aa-dfa0-48da-934c-80291354421d.mp4

**Found in Branch**
o3de_installer_2111_1

**Desktop/Device (please complete the following information):**

- Device: PC
- OS: Windows
- Version: 10 Pro
- CPU: AMD Ryzen 7 2700X
- GPU: AMD Radeon Pro Duo
- Memory: 16GB

**Additional context**

[error.zip](https://github.com/o3de/o3de/files/7626463/error.zip)

[error.log](https://github.com/o3de/o3de/files/7626464/error.log)

[Editor.log](https://github.com/o3de/o3de/files/7626465/Editor.log)

**Comment by SergeyAMZN** [comment]
From the log it seems the issue is the default material library we display in FBX settings hasn't been loaded yet. 
A bit strange since it is loaded with a blocking call when asset catalog is loaded on start up.

**Comment by AMZN-daimini** [comment]
Trying to reproduce this issue to get more information. May need a few fresh rebuilds/cache deletes depending on what's causing it to only happen on the first session :)

**Comment by AMZN-daimini** [comment]
Managed to get a repro. My log is a little different than what was reported, but actually sounds closer to @SergeyAMZN 's description.

```
 	PhysX.Editor.Gem.dll!AZStd::vector<Physics::MaterialFromAssetConfiguration,AZStd::allocator>::size() Line 434	C++
>	PhysX.Editor.Gem.dll!PhysX::Pipeline::MeshGroup::GetPhysicsMaterialNames() Line 872	C++
 	PhysX.Editor.Gem.dll!AZ::AttributeMemberFunction<AZStd::vector<AZStd::basic_string<char,AZStd::char_traits<char>,AZStd::allocator>,AZStd::allocator> (__cdecl PhysX::Pipeline::MeshGroup::*)(void)const>::Invoke(void * instance) Line 361	C++
 	EditorLib.dll!AZ::Internal::AttributeRead<AZStd::vector<AZStd::basic_string<char,AZStd::char_traits<char>,AZStd::allocator>,AZStd::allocator>,AZStd::vector<AZStd::basic_string<char,AZStd::char_traits<char>,AZStd::allocator>,AZStd::allocator>,void * &>(AZStd::vector<AZStd::basic_string<char,AZStd::char_traits<char>,AZStd::allocator>,AZStd::allocator> & value, AZ::Attribute * attr, void * & instance) Line 31	C++
 	EditorLib.dll!AZ::Internal::AttributeReader<AZStd::vector<AZStd::basic_string<char,AZStd::char_traits<char>,AZStd::allocator>,AZStd::allocator>,AZStd::vector<AZStd::basic_string<char,AZStd::char_traits<char>,AZStd::allocator>,AZStd::allocator>>::Read<void * &>(AZStd::vector<AZStd::basic_string<char,AZStd::char_traits<char>,AZStd::allocator>,AZStd::allocator> & value, AZ::Attribute * attr, void * & instance) Line 142	C++
 	EditorLib.dll!AZ::AttributeInvoker<void>::Read<AZStd::vector<AZStd::basic_string<char,AZStd::char_traits<char>,AZStd::allocator>,AZStd::allocator>,AZStd::vector<AZStd::basic_string<char,AZStd::char_traits<char>,AZStd::allocator>,AZStd::allocator>>(AZStd::vector<AZStd::basic_string<char,AZStd::char_traits<char>,AZStd::allocator>,AZStd::allocator> & value) Line 170	C++
 	EditorLib.dll!AzToolsFramework::PropertyComboBoxHandlerCommon<AZStd::basic_string<char,AZStd::char_traits<char>,AZStd::allocator>>::ConsumeAttribute(AzToolsFramework::PropertyStringComboBoxCtrl * GUI, unsigned int attrib, AZ::AttributeReader * attrValue, const char * debugName) Line 170	C++
 	EditorLib.dll!`AzToolsFramework::PropertyHandler_Internal<AzToolsFramework::PropertyStringComboBoxCtrl>::ConsumeAttributes_Internal'::`2'::<lambda_1>::operator()<AZStd::vector<AZStd::pair<unsigned int,AZ::Attribute *>,AZStd::allocator>>(const AZStd::vector<AZStd::pair<unsigned int,AZ::Attribute *>,AZStd::allocator> & attributes, const char * name) Line 50	C++
 	EditorLib.dll!AzToolsFramework::PropertyHandler_Internal<AzToolsFramework::PropertyStringComboBoxCtrl>::ConsumeAttributes_Internal(QWidget * widget, AzToolsFramework::InstanceDataNode * dataNode) Line 64	C++
 	SceneUI.dll!AzToolsFramework::ReflectedPropertyEditor::Impl::CreateEditorWidget(AzToolsFramework::PropertyRowWidget * pWidget) Line 646	C++
 	SceneUI.dll!AzToolsFramework::ReflectedPropertyEditor::Impl::ExpandChildren(AzToolsFramework::PropertyRowWidget * parentWidget, bool expand, bool checkIfChildrenShouldExpand) Line 671	C++
 	SceneUI.dll!AzToolsFramework::ReflectedPropertyEditor::Impl::ExpandChildren(AzToolsFramework::PropertyRowWidget * parentWidget, bool expand, bool checkIfChildrenShouldExpand) Line 674	C++
 	SceneUI.dll!AzToolsFramework::ReflectedPropertyEditor::Impl::UpdateExpansionState() Line 628	C++
 	SceneUI.dll!AzToolsFramework::ReflectedPropertyEditor::InvalidateAll(const char * filter) Line 898	C++
 	SceneUI.dll!AzToolsFramework::ReflectedPropertyEditor::DoRefresh() Line 2321	C++
 	SceneUI.dll!AzToolsFramework::ReflectedPropertyEditor::qt_static_metacall(QObject * _o, QMetaObject::Call _c, int _id, void * * _a) Line 110	C++
 	Qt5Cored.dll!QMetaCallEvent::placeMetaCall(QObject * object) Line 618	C++
 	Qt5Cored.dll!QObject::event(QEvent * e) Line 1314	C++
 	Qt5Widgetsd.dll!QWidget::event(QEvent * event) Line 9114	C++
 	Qt5Widgetsd.dll!QFrame::event(QEvent * e) Line 550	C++
 	Qt5Widgetsd.dll!QApplicationPrivate::notify_helper(QObject * receiver, QEvent * e) Line 3630	C++
 	Qt5Widgetsd.dll!QApplication::notify(QObject * receiver, QEvent * e) Line 3580	C++
 	Qt5Cored.dll!QCoreApplication::notifyInternal2(QObject * receiver, QEvent * event) Line 1063	C++
 	Qt5Cored.dll!QCoreApplication::sendEvent(QObject * receiver, QEvent * event) Line 1459	C++
 	Qt5Cored.dll!QCoreApplicationPrivate::sendPostedEvents(QObject * receiver, int event_type, QThreadData * data) Line 1817	C++
 	Qt5Cored.dll!QEventDispatcherWin32::sendPostedEvents() Line 1082	C++
 	[External Code]	
 	Qt5Cored.dll!QEventDispatcherWin32::processEvents(QFlags<enum QEventLoop::ProcessEventsFlag> flags) Line 530	C++
 	[External Code]	
 	Qt5Cored.dll!QEventLoop::processEvents(QFlags<enum QEventLoop::ProcessEventsFlag> flags) Line 140	C++
 	Qt5Cored.dll!QEventLoop::exec(QFlags<enum QEventLoop::ProcessEventsFlag> flags) Line 232	C++
 	Qt5Cored.dll!QCoreApplication::exec() Line 1371	C++
 	Qt5Guid.dll!QGuiApplication::exec() Line 1871	C++
 	Qt5Widgetsd.dll!QApplication::exec() Line 2823	C++
 	EditorLib.dll!CryEditMain(int argc, char * * argv) Line 4186	C++
 	Editor.exe!AZStd::Internal::INVOKE<int (__cdecl*&)(int,char * * const),int &,char * * &>(int(*)(int, char * *) & f, int & <args_0>, char * * & <args_1>) Line 209	C++
 	Editor.exe!AZStd::invoke<int (__cdecl*&)(int,char * * const),int &,char * * &>(int(*)(int, char * *) & f, int & <args_0>, char * * & <args_1>) Line 39	C++
 	Editor.exe!main(int argc, char * * argv) Line 32	C++
 	Editor.exe!WinMain(HINSTANCE__ * __formal, HINSTANCE__ * __formal, char * __formal, int __formal) Line 97	C++
 	[External Code]
```

![image](https://user-images.githubusercontent.com/82231674/144121618-06c33c52-2b61-4d92-8162-9d905e8b053d.png)
![image](https://user-images.githubusercontent.com/82231674/144121686-e088e33e-e566-4356-b326-033ffe458fc0.png)

In GetPhysicsMaterialNames:
```
AZStd::vector<AZStd::string> MeshGroup::GetPhysicsMaterialNames() const
{
    if (auto* physicsSystem = AZ::Interface<AzPhysics::SystemInterface>::Get())
    {
        if (const auto* physicsConfiguration = physicsSystem->GetConfiguration())
        {
            const auto& materials = physicsConfiguration->m_materialLibraryAsset->GetMaterialsData();

            AZStd::vector<AZStd::string> physicsMaterialNames;
            physicsMaterialNames.reserve(materials.size() + 1);

            physicsMaterialNames.emplace_back(Physics::DefaultPhysicsMaterialLabel);
            for (const auto& material : materials)
            {
                physicsMaterialNames.emplace_back(material.m_configuration.m_surfaceType);
            }

            return physicsMaterialNames;
        }
    }
    return {Physics::DefaultPhysicsMaterialLabel};
}
```
GetMaterialsData() returns an invalid reference. This matches Sergey's comment:
> From the log it seems the issue is the default material library we display in FBX settings hasn't been loaded yet.

**Comment by AMZN-daimini** [comment]
I have a small fix that seems to prevent the crash for me - I wasn't able to repro the issue again after introducing it.

This is not my area of expertise, so I'd ask for at least one review from the PhysX team before sending this in, just in case :)

**Comment by LB-MateuszSzweda** [comment]
Today it was verified that there is an existing workaround to avoid the issue on initial launch.
Running Asset Processor and letting it process all assets and only then running the Editor will not reproduce the issue.


So, running Editor (which launches asset processor by default) on initial launch - The issue reproduces

Running Asset Processor first and letting it process all assets before launching the Editor - The issue does not reproduce.

**Comment by moraaar** [comment]
I've investigated why physics material library is not ready and I just discovered something odd.

If I run the editor after Asset Processor has finished, the function `PhysXSystem::OnCatalogLoaded` is called correctly, which is the one that loads the physics material library if there is one assigned or triggers the logic to create the default one if there is none assigned to the PhysX configuration.

But if I run the editor while there is no cache folder yet, the asset processor will open, start processing, editor will hold until SOME (not all) assets are processed, then open the editor, but the function `PhysXSystem::OnCatalogLoaded` will NEVER be called, even after Asset Processor fully finished.

Since `PhysXSystem::OnCatalogLoaded` never gets called in that scenario (which is the one from this bug), then for a new project the default physics material library will never be created and it crashes the first time some code asks for it (fbx setting in the case of this bug). For an existing project (for example `AutomatedTesting`) the physics material library assigned will not be loaded and it will just return "Default Material", which would be incorrect.

So to properly fix this bug we have to:

1. Mark .physmaterial assets as criticals so the editor doesn't open without Asset Processor processing them. 
2. Find out why OnCatalogLoaded is not being broadcast when the editor is launched before processing any asset.

**Comment by moraaar** [comment]
PR #6050 fixed the crash, but physx materials shown in FBX Settings are still not behaving as it should (it only shows default material, instead of all the materials from the library). This is due to a larger issue discovered while investigating this:

#6093

**Comment by moraaar** [comment]
Also, PR #6050 was merged into development branch.

@forhalle @yuyihsu should this be cherry picked into stabilization/2111RTE branch too?

**Comment by forhalle** [comment]
@moraaar  - Yes, PR #6050, which is the fix for #6027, should go into stabilization/2111RTE.

**Comment by yuyihsu** [comment]
Sorry, I accidentally removed the "feature/physics" label yesterday, adding it back.

**Comment by moraaar** [comment]
> @moraaar - Yes, PR #6050, which is the fix for #6027, should go into stabilization/2111RTE.

Thanks, I will.

**Comment by AMZN-daimini** [comment]
@moraaar can you take care of the cherry-pick? If so, I would assign this issue to you.
Else let me know :D

**Comment by moraaar** [comment]
> @moraaar can you take care of the cherry-pick? If so, I would assign this issue to you. Else let me know :D

Done: https://github.com/o3de/o3de/commit/dc8f9dc233b3819a5c4e6c2e0de273e635213c60

**Comment by LB-MateuszSzweda** [comment]
Verified the issue as fixed on Stabilization/2111RTE (fb36145) and Development (df8326a).
The Editor no longer crashes when PhysX meshes are picked for .fbx assets and in updating, material libraries are also functional.

However, there seems to be a residual minor still as .fbx assets in Asset Browser still have a placeholder icon and will not have preview available upon the first launch of the Editor. (this resolves when the Editor is restarted) 
As this is not a critical issue I'll have it logged separately instead of reopening the above.
